src/java/org/pantsbuild/tools/junit/impl/ShutdownListener.java

```diff
@@ -3,71 +3,88 @@
 
 package org.pantsbuild.tools.junit.impl;
 
-import java.io.PrintStream;
+import java.util.concurrent.ConcurrentHashMap;
 import org.junit.runner.Description;
 import org.junit.runner.Result;
 import org.junit.runner.notification.Failure;
 import org.junit.runner.notification.RunListener;
 
 /**
- * A listener that keeps track of the current test state with its own result class so it can print
+ * A listener that keeps track of the current test state with its own result class so it can record
  * the state of tests being run if there is unexpected exit during the tests.
  */
-public class ShutdownListener extends ConsoleListener {
+public class ShutdownListener extends RunListener {
   private final Result result = new Result();
   private final RunListener resultListener = result.createListener();
-  private Description currentTestDescription;
+  private ConcurrentHashMap.KeySetView<Description, Boolean> currentDescriptions =
+    ConcurrentHashMap.newKeySet();
+  private RunListener underlying;
+  private static Throwable abnormalExit =
```

Comment:
Constructing a static exception like this will result in a non-useful stack trace. Is it necessary for this to be static, or can it be created at the relevant location?

Reply:
It's not necessary--it results in practically the same stacktrace, which isn't very useful since it's the stacktrace of the shutdown hook, taking place mostly in this file and in Thread--but I'll switch away from the somewhat stranger static exception